src/devices/src/virtio/block/device.rs

```diff
@@ -332,886 +332,641 @@ impl VirtioDevice for Block {
+
+            invoke_handler_for_queue_event(&mut block);
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 1);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_IOERR
+            );
+        }
+    }
+
+    #[test]
+    fn test_unsupported_request_type() {
+        let mut block = default_block();
+        let mem = block.mem.clone();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, 16);
+        block.set_queue(0, vq.create_queue());
+        block.set_device_activated(true);
+        initialize_virtqueue(&vq);
+
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        // Set sector to 0.
+        mem.write_obj_at_addr::<u64>(0, GuestAddress(0x1000 + 8))
+            .unwrap();
+        // Currently only 0, 1, 4, 8 are supported.
+        // Generate an unsupported request.
+        mem.write_obj_at_addr::<u32>(16, GuestAddress(0x1000))
+            .unwrap();
+
+        invoke_handler_for_queue_event(&mut block);
+
+        assert_eq!(vq.used.idx.get(), 1);
+        assert_eq!(vq.used.ring[0].get().id, 0);
+        assert_eq!(vq.used.ring[0].get().len, 1);
+        assert_eq!(
+            mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+            VIRTIO_BLK_S_UNSUPP
+        );
+    }
+
+    #[test]
+    fn test_read_write() {
+        let mut block = default_block();
+        let mem = block.mem.clone();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, 16);
+        block.set_queue(0, vq.create_queue());
+        block.set_device_activated(true);
+        initialize_virtqueue(&vq);
+
+        let data_addr = GuestAddress(vq.dtable[1].addr.get());
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        // Write.
+        {
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_OUT, GuestAddress(0x1000))
+                .unwrap();
+            // Make data read only, 8 bytes in len, and set the actual value to be written.
+            vq.dtable[1].flags.set(VIRTQ_DESC_F_NEXT);
+            vq.dtable[1].len.set(8);
+            mem.write_obj_at_addr::<u64>(123_456_789, data_addr)
+                .unwrap();
+
+            check_metric_after_block!(
+                &METRICS.block.write_count,
+                1,
+                invoke_handler_for_queue_event(&mut block)
+            );
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 0);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_OK
+            );
+        }
+
+        // Read.
+        {
+            vq.used.idx.set(0);
+            block.set_queue(0, vq.create_queue());
+
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_IN, GuestAddress(0x1000))
```

Comment:
Use a variable here to make the `GuestAddress(0x1000)` more meaningful.

Reply:
Done.